### PR TITLE
table: add bypassing cache when generating view updates

### LINKED ISSUE #4615: restart_node_with_resharding nemesis seastar::semaphore_timed_out (Semaphore timedout)
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 3.1.0.rc2-0.20190624.433cb93f7
Cluster size: 5
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0ac07c9630d9e4eea
(Ireland)

Platform (physical/VM/cloud instance type/docker): i3.4xlarge

During the nemesis restart_node_with_resharding, the target node, all of the shards returned the error "semaphore_timed_out" from all of its shards.
```
Jun 26 13:36:25 ip-10-0-167-66.eu-west-1.compute.internal scylla.bin[17945]:  [shard 6] storage_proxy - Failed to apply mutation from 10.0.132.172#6: seastar::semaphore_timed_out (Semaphore timedout)
```
[journal.log](https://github.com/scylladb/scylla/files/3330726/journal.log)

### PR DESCRIPTION
**Pull request by psarna**
Generating view updates can be a source of high read amplification,
since updates may need to perform read-before-write.
After ensuring that readers participating in this read-before-write
process use `bypass_cache`, the read amplification from a test case:

1. Creating an index
  CREATE INDEX index1  ON myks2.standard1 ("C1")
2. Running cassandra-stress in order to generate view updates
cassandra-stress write no-warmup n=1000000 cl=ONE -schema \
  'replication(factor=2) compaction(strategy=LeveledCompactionStrategy)' \
  keyspace=myks2 -pop seq=4000000..8000000 -rate threads=100 -errors
  skip-read-validation -node 127.0.0.1

... dropped to 1.5GB, down from quite astonishing 36GB (sic!)
from before the change.

Fixes #4615

### CODE REVIEW

**Comment by psarna** [comment]
Note: I added this pull request so it can be referenced, but before merging it I'd like to make sure I understand why the results are so vast.

**Comment by denesb** [comment]
How do you measure this read amplification?

**Comment by psarna** [comment]
@bdenes the whole story is here: https://github.com/scylladb/scylla-enterprise/issues/1028

tl;dr: from io queue metrics, my script looks like this:
```bash
#!/bin/bash

HOST=${1:-127.0.0.2}

curl -slient ${HOST}:9180/metrics | grep scylla_io_queue_total_bytes| grep -v HELP | grep -v TYPE | awk '{ foo = $2 / 1024 / 1024 / 1024 ; print $1 "\t" foo " GiB" }'
```
the data size is less than 1GB, so 36GB reads are way over the top, that's how the whole issue started.

**Comment by avikivity** [comment]
Suppose the same row is updated several times in a short time. Won't this cause a regression, because those reads-before-writes will miss?

**Comment by avikivity** [comment]
I don't understand how bypass cache reduces disk reads. It should increase them!

**Comment by psarna** [comment]
@avikivity I don't understand that at all either, and yet that's what both metrics and `iotop` confirms... it definitely helped a lot with the read ratio of a write-intensive workload, when a materialized view was present.

**Comment by avikivity** [comment]
Well, we should understand it before committing anything.

**Comment by psarna** [comment]
Yes, certainly, I already added this note as a first comment in this pull request - perhaps I'll also negatively review it myself so nobody is tempted to click on the merge button.

**Comment by psarna** [comment]
I can't review it myself, even in a negative way - thanks github!

**Comment by psarna** [comment]
As it turned out in the discussion, the real culprit was an implicit `fwd_mr::yes` parameter passed to a constructor of a reader for mv's read-before-write, which turns single-partition reads into scans, which in turn results in read amplification.
The solution is to make sure that readers created for reading single partitions for read-before-write are not accidentally turned into scanning readers by erroneously enabling fast forwarding in them. I'll push a separate pull request with a fix to avoid confusion.
